### PR TITLE
Support 3D-MoE LoRA export for shared-outer routed-expert adapters

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Support vLLM 3D-MoE LoRA export (`stack_3d_moe`) for shared-outer routed-expert adapters

# Changelog

- New `stack_3d_moe: bool` flag on `export_adapter_weights` / `stream_adapter_weights_megatron_to_hf`. When set, shared-outer routed-expert LoRA is emitted as the two stacked 3D tensors vLLM's `FusedMoE3DWithLoRA` consumer looks up (`...experts.base_layer.lora_{A,B}.weight` for gate_up_proj, bare `...experts.lora_{A,B}.weight` for down_proj), with axis order `lora_A=(E, r, in)` and `lora_B=(out, r, E)`. This is the 3D-MoE analogue of the existing 2D `expand_shared_outer` path and takes precedence over it for shared-outer adapters. Without this, vLLM 3D-MoE models (`is_3d_moe_weight=True`, e.g. Qwen3.5/3.6 VLM) silently drop every routed-expert LoRA → garbled output, reward=0.
- 3 new tests — `test_stack_3d_moe_gate_up_emits_base_layer_3d`, `test_stack_3d_moe_down_emits_bare_3d`, `test_stack_3d_moe_matches_vllm_consumer_layout`.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: src/megatron/bridge/models/conversion/peft_bridge.py [hunk 1]
```diff
@@ -1045,11 +1054,33 @@ def _stream_shared_outer_adapter_weights(
...
+                megatron_model,
+                mapping_registry,
+                adapter_task,
+                linear_in_tensor,
+                linear_out_tensor,
+                num_moe_experts,
+                cpu,
+                is_expert,
+            )
+            return
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/peft_bridge.py:1082]: _stream_shared_outer_adapter_weights is only entered via if is_shared_outer_lora: (L897), where is_shared_outer_lora = is_grouped_expert and ... and is_grouped_expert = is_expert and .... So is_expert is always True inside this function — the proposed raise ValueError is unreachable dead code. The sub-method already has a graceful guard at L1189 (if not base_hf_weight_names: return). Adding the guard would make the code worse — crashing where it can't crash.

File: src/megatron/bridge/models/conversion/peft_bridge.py [hunk 2]
```diff
@@ -1114,6 +1145,96 @@ def _stream_shared_outer_adapter_weights(
...
+        # all global experts yields the (E, out, r) tensor vLLM expects, already
+        # carrying gate+up fused on the output dim for FC1.
+        gathered_out = self._gather_expert_adapter_weight(linear_out_tensor)
+        linear_out_all = torch.stack(
+            [
+                self._select_expert_adapter_weight(linear_out_tensor, gathered_out, i, num_moe_experts)
+                for i in range(num_moe_experts)
+            ],
+            dim=0,
+        )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply src/megatron/bridge/models/conversion/peft_bridge.py]: When EP=1, _gather_expert_adapter_weight returns None, and the Python loop at L1223-1229 slices weight[i] then torch.stacks them back — a full redundant reconstruction of the original 3D tensor. For 256 experts this is avoidable memory/CPU overhead during export. Added fast path: when gathered_out is None and linear_out_tensor.ndim > 2, use linear_out_tensor directly.

File: tests/unit_tests/models/test_adapter_export.py [hunk 3]
```diff
@@ -1695,3 +1695,168 @@ def test_per_expert_side_unaffected_by_flag(self, _mock_ep):
...
+        mapping = MagicMock()
+
+        def lookup(name):
+            idx = name.rsplit(".weight", 1)[1]
+            return SimpleNamespace(
+                hf_param=f"model.layers.0.mlp.experts.{idx}.gate_proj.weight"
+            )
+
+        mapping.megatron_to_hf_lookup.side_effect = lookup
+        return mapping
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply tests/unit_tests/models/test_adapter_export.py:1722]: suggested diff was buggy — it replaced hf_param=str with hf_param=dict without accounting for the idx parsing logic. The real GatedMLPMapping has hf_param={"gate": "...gate_proj.weight", "up": "...up_proj.weight"} as a dict. Fixed to return a dict with both gate and up keys, so _get_base_hf_param_names_for_adapter yields both projections and the is_gate_up detection path is actually exercised.

**Review by yaoyu-33 — DISMISSED** [review]

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5939/commits/58490f5ca01ac3a43fc49a63418abf87a09ac11e

**Comment by liding-nv** [comment]
@HollowMan6 could u fix the unit test so we can re-run it

**Comment by HollowMan6** [comment]
Thank you @liding-nv for reminding, now I've just pushed the commit to fix the failed test case.

**Review by liding-nv — APPROVED** [review]
